pkg/blockbuilder/scheduler/priority_queue.go

```diff
@@ -4,91 +4,158 @@ import (
-type PriorityQueue[T any] struct {
-	h *priorityHeap[T]
+// PriorityQueue is a generic priority queue with constant time lookups.
+type PriorityQueue[K comparable, V any] struct {
+	h   *priorityHeap[V]
+	m   map[K]*item[V] // Map for constant time lookups
+	key func(V) K      // Function to extract key from value
+}
+
+// item represents an item in the priority queue with its index
+type item[V any] struct {
+	value V
+	index int
 }
 
 // NewPriorityQueue creates a new priority queue.
-func NewPriorityQueue[T any](less func(T, T) bool) *PriorityQueue[T] {
-	h := &priorityHeap[T]{
+func NewPriorityQueue[K comparable, V any](less func(V, V) bool, key func(V) K) *PriorityQueue[K, V] {
+	h := &priorityHeap[V]{
 		less: less,
-		heap: make([]T, 0),
+		heap: make([]*item[V], 0),
+		idx:  make(map[int]*item[V]),
 	}
 	heap.Init(h)
-	return &PriorityQueue[T]{h: h}
+	return &PriorityQueue[K, V]{
+		h:   h,
+		m:   make(map[K]*item[V]),
+		key: key,
+	}
 }
 
 // Push adds an element to the queue.
-func (pq *PriorityQueue[T]) Push(v T) {
-	heap.Push(pq.h, v)
+func (pq *PriorityQueue[K, V]) Push(v V) {
+	k := pq.key(v)
+	if existing, ok := pq.m[k]; ok {
+		// Update existing item's value and fix heap
+		existing.value = v
+		heap.Fix(pq.h, existing.index)
+		return
+	}
+
+	// Add new item
+	idx := pq.h.Len()
+	it := &item[V]{value: v, index: idx}
+	pq.m[k] = it
+	heap.Push(pq.h, it)
 }
 
 // Pop removes and returns the element with the highest priority from the queue.
-func (pq *PriorityQueue[T]) Pop() (T, bool) {
+func (pq *PriorityQueue[K, V]) Pop() (V, bool) {
 	if pq.Len() == 0 {
-		var zero T
+		var zero V
 		return zero, false
 	}
-	return heap.Pop(pq.h).(T), true
+	it := heap.Pop(pq.h).(*item[V])
+	delete(pq.m, pq.key(it.value))
+	return it.value, true
+}
+
+// Lookup returns the item with the given key if it exists.
+func (pq *PriorityQueue[K, V]) Lookup(k K) (V, bool) {
+	if it, ok := pq.m[k]; ok {
+		return it.value, true
+	}
+	var zero V
+	return zero, false
+}
+
+// Remove removes and returns the item with the given key if it exists.
+func (pq *PriorityQueue[K, V]) Remove(k K) (V, bool) {
+	it, ok := pq.m[k]
+	if !ok {
+		var zero V
+		return zero, false
+	}
+	heap.Remove(pq.h, it.index)
+	delete(pq.m, k)
+	return it.value, true
+}
+
+// UpdatePriority updates the priority of an item and reorders the queue.
+func (pq *PriorityQueue[K, V]) UpdatePriority(k K, v V) bool {
+	if it, ok := pq.m[k]; ok {
+		it.value = v
+		heap.Fix(pq.h, it.index)
+		return true
+	}
+	return false
 }
 
 // Len returns the number of elements in the queue.
-func (pq *PriorityQueue[T]) Len() int {
+func (pq *PriorityQueue[K, V]) Len() int {
 	return pq.h.Len()
 }
 
 // priorityHeap is the internal heap implementation that satisfies heap.Interface.
-type priorityHeap[T any] struct {
-	less func(T, T) bool
-	heap []T
+type priorityHeap[V any] struct {
+	less func(V, V) bool
+	heap []*item[V]
+	idx  map[int]*item[V] // Maps index to item for efficient updates
```

Comment:
do we use this for lookups? i don't see any references

Reply:
good call 👍